Added TraversabilityGridVisualization.

JanWehrmann opened the pull request:
TraversabilityGridVisualization is heavily based
on GridMapVisualization and MLSMapVisualization.

Creates a plane and applies a texture. Color for
each cell is chosen depending on the
traversability.

Colors for UNKOWN, OBSTACLE and the minimum and
maximum traversability can be chosen by the user.

saarnold commented on viz/TraversabilityGridVisualization.cpp:
```diff
@@ -0,0 +1,329 @@
+{
+    maps::grid::TraversabilityGrid traversabilityGrid;
+
+    Data(const maps::grid::TraversabilityGrid traversabilityGrid)
```
TraversabilityGrid should be a reference here.

chhtz replied:
Actually, this is also wrong in `MLSMapVisualization`. And I'm not sure, if
    ```Data(maps::grid::TraversabilityGrid tG) : traversabilityGrid(std::move(tG)) {}```
might actually be better (I guess overall, it does not matter too much).

saarnold replied:
True, it should be fixed for the MLS visualization as well.
What I don't like in case of the move operation is that `tG` would be unspecified in the constructor scope. To whether it is more efficient I can't tell.

saarnold commented on viz/TraversabilityGridVisualization.cpp:
```diff
@@ -0,0 +1,329 @@
+
+using namespace ::vizkit3d;
+
+struct TraversabilityGridVisualization::Data
```
I guess you added all the visualization functions to the Data struct since it was done in the MLS visualization equally.
The reason for this, in case of the MLS, was that it is templated in order to support all different kinds of MLS maps. This is not necessary here. Please move the functions to the TraversabilityGridVisualization class. You can use the Data struct to hold the TraversabilityGrid but it can as well be a member of TraversabilityGridVisualization.

JanWehrmann replied:
Ahh well. I left out DataHold since it wasn't needed without any templates but didn't question wether or not I actually needed Data in that case. Thanks!

chhtz replied:
Right, here it is entirely unnecessary.

saarnold commented on viz/TraversabilityGridVisualization.cpp:
```diff
@@ -0,0 +1,321 @@
+    osg::ref_ptr<osg::Geode> geode = new osg::Geode();
+    group->addChild(geode.get());
+
+    return group.release();
```
There is no need to release the object here, since you want to return a `ref_ptr`.

saarnold commented on viz/TraversabilityGridVisualization.cpp:
```diff
@@ -0,0 +1,321 @@
+void TraversabilityGridVisualization::updateMainNode(osg::Node* node)
+{
+    osg::ref_ptr<osg::Group> group = node->asGroup();
+    group->removeChildren(0, group->getNumChildren());
```
Ideally `createMainNode` creates as much of the osg graph as possible and `updateMainNode` only modifies existing objects and removes and creates new drawables when necessary.
Not that you can hold certain objects of the graph as members or your class if you need to identify them (With the constraint that they must only be modified when `updateMainNode` is called).

saarnold commented on test/viz/test_viz_TraversabilityGrid.cpp:
```diff
@@ -0,0 +1,108 @@
+    }
+}
+
+TraversabilityGrid sinesMlsToTravGrid(Vector2ui numCells, Vector2d resolution)
```
Eigen types should always be referenced parameters.

saarnold commented on viz/TraversabilityGridVisualization.cpp:
```diff
@@ -0,0 +1,321 @@
+        state->setTextureAttributeAndModes(0, tex.get());
+}
+
+void vizkit3d::TraversabilityGridVisualization::visualizeMapExtents(osg::Group* group) const
```
Ne need to use a raw pointer here, use a reference instead.

saarnold commented on viz/ExtentsRectangle.hpp:
```diff
@@ -65,6 +67,16 @@ class ExtentsRectangle : public osg::Geode
         ss->setAttribute( new osg::LineWidth( 3.0 ) );
     }
+
+     void update(Eigen::Vector2d min, Eigen::Vector2d max)
```
Please use const references

saarnold commented on viz/TraversabilityGridVisualization.cpp:
```diff
@@ -0,0 +1,325 @@
+
+    if (showMapExtents)
+    {
+        if (numChildren <= 1)
```
Please note that you can ask group if it has a specific child. See [containsNode](http://camlosg.sourceforge.net/osg/classosg_1_1Group.html#5b7d6aa9a106792464684790f6b377e4).

saarnold commented on viz/TraversabilityGridVisualization.cpp:
```diff
@@ -0,0 +1,325 @@
+
+        osg::ref_ptr<osg::Image> image = createTextureImage();
+
+        osg::ref_ptr<osg::Texture2D> tex = new osg::Texture2D(image.get());
```
Does the Texture2D has to be recreated to change it's properties?

JanWehrmann committed:
Added TraversabilityGridVisualization.

Creates a plane and applies a texture. Color for
each cell is chosen depending on the
traversability.

Colors for UNKOWN, OBSTACLE and the minimum and
maximum traversability can be chosen by the user.

Includes all necessary changes to CMakeLists etc.
as well as some tests.
